pkg/types/openstack/validation/machinepool_test.go

```diff
@@ -47,29 +63,91 @@ func TestValidateMachinePool(t *testing.T) {
 	for _, tc := range [...]struct {
 		name        string
 		machinePool *openstack.MachinePool
+		role        string
 		checks      []checkFunc
 	}{
 		{
 			"empty",
 			testMachinePool(),
+			"default",
 			check(noError),
 		},
 		{
 			"with valid server group policy",
 			testMachinePool(withServerGroupPolicy("anti-affinity")),
+			"default",
 			check(noError),
 		},
 		{
 			"with invalid server group policy",
 			testMachinePool(withServerGroupPolicy("anti-gravity")),
+			"default",
 			check(
 				someErrorType(field.ErrorTypeNotSupported),
 				exactlyNErrors(1),
 			),
 		},
+		{
+			"failure domains",
+			testMachinePool(withFailureDomain(openstack.FailureDomain{
+				PortTargets: []openstack.NamedPortTarget{{ID: "one"}, {ID: "two"}},
+			})),
+			"master",
+			check(noError),
+		},
+		{
+			"failure domains on worker",
+			testMachinePool(withFailureDomain(openstack.FailureDomain{
+				PortTargets: []openstack.NamedPortTarget{{ID: "one"}, {ID: "two"}},
+			})),
+			"worker",
+			check(
+				someErrorType(field.ErrorTypeForbidden),
+				exactlyNErrors(1),
+			),
+		},
+		{
+			"failure domains, duplicate portTarget ID",
+			testMachinePool(withFailureDomain(openstack.FailureDomain{
+				PortTargets: []openstack.NamedPortTarget{{ID: "one"}, {ID: "one"}},
+			})),
+			"master",
+			check(
+				someErrorType(field.ErrorTypeDuplicate),
+				exactlyNErrors(1),
+			),
+		},
+		{
+			"failure domains together with zones",
+			testMachinePool(
+				withZones(),
+				withFailureDomain(openstack.FailureDomain{
+					PortTargets: []openstack.NamedPortTarget{{ID: "one"}, {ID: "two"}},
+				}),
+			),
+			"worker",
```

Comment:
Should be master, otherwise we can't differentiate with the `failure domains on worker` test.